ARC: MWDT: drop redundant stack checking

evgeniy-paltsev opened the pull request:
MWDT toolchain has Stackcheck_alloca option enabled by default. So it adds extra stack checking in addition to Zephyr's stack checking.

As it is completely redundant let's drop it.

ruuddw commented:
The check is redundant indeed and can be disabled in MetaWare.
But, I'm not fully happy with repeating this generic option for all different boards/SoCs that we support.
Also, just noted that only for the nSIM boards we have MetaWare build flags defined.
Can the "-Hoff=Stackcheck_alloca" be moved to a common file? As this applies to all ARC targets, move it to the MetaWare compiler support?

ruuddw reviewed (changes requested):
See comments, please factor out the compiler/linker option so it doesn't have to be repeated for every ARC SoC.

Commit:
ARC: MWDT: drop redundant stack checking

MWDT toolchain has Stackcheck_alloca option enabled by default.
So it adds stack checking in addition to Zephyr's stack checking.

As it is completely redundant let's drop it.

Signed-off-by: Eugeniy Paltsev <Eugeniy.Paltsev@synopsys.com>

evgeniy-paltsev commented:
@ruuddw Good point, fixed.

andrewboie commented:
Question: what does `Stackcheck_alloca` actually do?

CONFIG_HW_STACK_PROTECTION ensures we don't go past the bounds of the entire stack buffer. But it is not stack canaries or otherwise a mechanism to check the validity of individual stack frames, which is typically what these compiler features do.

I'd like a clearer description of what functionality is duplicated.

evgeniy-paltsev commented:
@andrewboie 
`Stackcheck_alloca` adds runtime check [on `alloca()` function call] that stack pointer still falls within the bounds of the entire stack buffer. It doesn't check validity of individual stack frames, etc...

andrewboie commented:
> @andrewboie
> `Stackcheck_alloca` adds runtime check [on `alloca()` function call] that stack pointer still falls within the bounds of the entire stack buffer. It doesn't check validity of individual stack frames, etc...

Interesting. How does this mechanism know what the stack buffer boundaries are? I see no hooks to query the kernel for this information.

evgeniy-paltsev commented:
Stack boundaries must be provided via `_fstack` and `_estack` variables. As of today this stack checking is suppressed as we provide whole RAM region boundaries instead of stack buffer boundaries (you can check it in dropped hunk from ARC linker script)

ruuddw commented:
Btw, MetaWare supports function-level canary checking as well - as the same _stackCheckHandler and __stack_chk_guard are used as for gcc, no special handling is required.
